### PR TITLE
CheckRightService_AddRightsForUser

### PR DESCRIPTION
**Pull request by ifelseelif**
Task
- create endpoint
Request: userId, rights ids list
Response: bool

Steps :
- Added validator for request
- Added command for adding rights for user
- Updated repository
- Added unit tests
- Added table DbRightUsers

### CODE REVIEW

File: DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace LT.DigitalOffice.CheckRightsService.Validator
+{
+    public class AddRightsForUserValidator : AbstractValidator<RightsForUserRequest>
+    {
+        public AddRightsForUserValidator()
+        {
+            RuleFor(rights => rights.UserId)
+                .NotEmpty()
+                .WithMessage("User id is empty.");
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs]: Remove it message.

File: DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            
+            RuleFor(rights => rights.RightsId)
+                .NotEmpty()
+                .WithMessage("No rights to add.");
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs]: Remove it message.

File: DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs (hunk 1 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs]: Why?

Comment (KostinP) [reply DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs]: https://digitaloffice-group.slack.com/archives/C016H4Z2LGL/p1595663778001900

Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs]: I changed on WithName

File: DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs (hunk 2 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Validator/AddRightsForUserValidator.cs]: I changed on WithName

**Review by MihailKulikov — COMMENTED** [review]
Maybe there need to be a check if a user, who is adding the right, has rights for this action.

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿using LT.DigitalOffice.CheckRightsService.RestRequests;
+
+namespace LT.DigitalOffice.CheckRightsService.Commands.Interfaces
+{
+    public interface IAddRightsForUserCommand
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs:8]: Add XML.

File: DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs [hunk 4]
```diff
@@ -15,5 +16,14 @@ public List<Right> GetRightsList([FromServices] IGetRightsListCommand command)
         {
             return command.Execute();
         }
+        
+        [Route("AddRightsForUser")]
+        [HttpPost]
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs]: Can be combined.

File: DigitalOffice/Services/CheckRightsService/Database/Entities/DbRightUser.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace LT.DigitalOffice.CheckRightsService.Database.Entities
+{
+    public class DbRightUser
+    {
+        public Guid UserId { get; set; }
+
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Database/Entities/DbRightUser.cs:10]: Why is there an empty line?

**Review by KostinP — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs (hunk 4 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs]: fixed

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs (hunk 3 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs:8]: added

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿using LT.DigitalOffice.CheckRightsService.RestRequests;
+
+namespace LT.DigitalOffice.CheckRightsService.Commands.Interfaces
+{
+    /// <summary>
+    /// Add rights for user
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs]: We write comments with a dot at the end.

**Review by KostinP — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs (hunk 6 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs]: fixed

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace LT.DigitalOffice.CheckRightsService.Commands.Interfaces
+{
+    /// <summary>
+    /// Add rights for user.
+    /// </summary>
+    public interface IAddRightsForUserCommand
+    {
+        /// <summary>
+        /// Add rights for user
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs]: Add dot.

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+        /// </summary>
+        /// <param name="request">Request with rights and user id</param>
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs]: Add dot.

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        /// <returns>Return true if successfully else return false</returns>
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IAddRightsForUserCommand.cs]: Add dot.

File: DigitalOffice/Services/CheckRightsService/RestRequests/RightsForUserRequest.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace LT.DigitalOffice.CheckRightsService.RestRequests
+{
+    public class RightsForUserRequest
+    {
+        public Guid UserId { get; set; }
+        public List<int> RightsId { get; set; }
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/RestRequests/RightsForUserRequest.cs]: List to IEnumerable.

**Review by KostinP — CHANGES_REQUESTED** [review]

**Review by KostinP — APPROVED** [review]

**Comment by KostinP** [comment]
But you just need to take into account Misha's comment (I think you can discuss this today on the call)
Similarly, in my pull request about deleting rights

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs [hunk 11]
```diff
@@ -19,5 +21,30 @@ public List<DbRight> GetRightsList()
         {
             return dbContext.Rights.ToList();
         }
+
+        public bool AddRightsToUser(RightsForUserRequest request)
+        {
+            foreach (var rightId in request.RightsId)
+            {
+                var dbRight = dbContext.Rights.FirstOrDefault(right => right.Id == rightId);
+                if (dbRight == null)
```
Comment (djino97) [inline DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: add emty line

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs [hunk 12]
```diff
@@ -19,5 +21,30 @@ public List<DbRight> GetRightsList()
...
+                    throw new Exception("Right doesn't exist.");
+                }
+
+                dbRight.UserIds.Add(new DbRightUser
+                {
+                    UserId = request.UserId,
+                    RightId = rightId,
+                });
+
+                dbContext.Rights.Update(dbRight);
```
Comment (djino97) [inline DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: Move `dbContext.Rights.Update(dbRight);` from block `foreach`.

File: DigitalOffice/Tests/CheckRightsServiceTests/Commands/AddRightsForUserCommandTests.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+using LT.DigitalOffice.CheckRightsService.Commands.Interfaces;
+using LT.DigitalOffice.CheckRightsService.Repositories.Interfaces;
+using LT.DigitalOffice.CheckRightsService.RestRequests;
+using Moq;
+using NUnit.Framework;
+
+namespace LT.DigitalOffice.CheckRightsServiceUnitTests.Commands
+{
+    public class AddRightsForUserCommandTests
+    {
```
Comment (djino97) [inline DigitalOffice/Tests/CheckRightsServiceTests/Commands/AddRightsForUserCommandTests.cs:15]: And if repository throw exception? Where test?

File: DigitalOffice/Tests/CheckRightsServiceTests/Validators/AddRightsForUserValidatorTests.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+using NUnit.Framework;
+
+namespace LT.DigitalOffice.CheckRightsServiceUnitTests.Validators
+{
+    public class AddRightsForUserValidatorTests
+    {
+        private IValidator<RightsForUserRequest> validator;
+
+        [SetUp]
+        public void Initialize()
```
Comment (djino97) [inline DigitalOffice/Tests/CheckRightsServiceTests/Validators/AddRightsForUserValidatorTests.cs]: SetUp

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs [hunk 15]
```diff
@@ -19,5 +21,30 @@ public List<DbRight> GetRightsList()
...
+            foreach (var rightId in request.RightsId)
+            {
+                var dbRight = dbContext.Rights.FirstOrDefault(right => right.Id == rightId);
+                if (dbRight == null)
+                {
+                    //TODO change on custom exception
+                    throw new Exception("Right doesn't exist.");
+                }
+
+                dbRight.UserIds.Add(new DbRightUser
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: Why not AddRange?

File: DigitalOffice/Services/CheckRightsService/RestRequests/RightsForUserRequest.cs [hunk 16]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace LT.DigitalOffice.CheckRightsService.RestRequests
+{
+    public class RightsForUserRequest
+    {
+        public Guid UserId { get; set; }
+        public IEnumerable<int> RightsId { get; set; }
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/RestRequests/RightsForUserRequest.cs]: RightIds?

File: DigitalOffice/Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs [hunk 17]
```diff
@@ -11,35 +13,32 @@ namespace LT.DigitalOffice.CheckRightsServiceUnitTests.Repositories
...
         }
 
         [TearDown]
         public void Clear()
         {
-            if (dbContext.Database.IsInMemory())
-            {
-                dbContext.Database.EnsureDeleted();
-            }
+            if (dbContext.Database.IsInMemory()) dbContext.Database.EnsureDeleted();
```
Comment (KostinP) [inline DigitalOffice/Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs]: Why?

File: DigitalOffice/Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs [hunk 18]
```diff
@@ -11,35 +13,32 @@ namespace LT.DigitalOffice.CheckRightsServiceUnitTests.Repositories
...
         }
 
         [Test]
         public void ShouldGetRightsListWhenDbIsNotEmpty()
         {
-            dbRight = new DbRight { Id = 0, Name = "Right", Description = "Allows you everything" };
+            dbRight = new DbRight {Id = 0, Name = "Right", Description = "Allows you everything"};
```
Comment (KostinP) [inline DigitalOffice/Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs]: Why?

**Review by KostinP — CHANGES_REQUESTED** [review]

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs [hunk 19]
```diff
@@ -19,5 +21,30 @@ public List<DbRight> GetRightsList()
...
+        public bool AddRightsToUser(RightsForUserRequest request)
+        {
+            foreach (var rightId in request.RightsId)
+            {
+                var dbRight = dbContext.Rights.FirstOrDefault(right => right.Id == rightId);
+                if (dbRight == null)
+                {
+                    //TODO change on custom exception
+                    throw new Exception("Right doesn't exist.");
+                }
```
Comment (djino97) [inline DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: NullReferenceException

File: DigitalOffice/Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs (hunk 18 above)
Comment (ifelseelif) [reply DigitalOffice/Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs]: it's Resharper code cleanup. fixed

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs (hunk 19 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: Do you mean change on NullReferenceException?

File: DigitalOffice/Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs (hunk 17 above)
Comment (ifelseelif) [reply DigitalOffice/Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs]: fixed

File: DigitalOffice/Services/CheckRightsService/RestRequests/RightsForUserRequest.cs (hunk 16 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/RestRequests/RightsForUserRequest.cs]: fixed

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs (hunk 11 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: added

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs (hunk 12 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: Why? dbRight change in the block foreach

File: DigitalOffice/Tests/CheckRightsServiceTests/Commands/AddRightsForUserCommandTests.cs (hunk 13 above)
Comment (ifelseelif) [reply DigitalOffice/Tests/CheckRightsServiceTests/Commands/AddRightsForUserCommandTests.cs:15]: added

File: DigitalOffice/Tests/CheckRightsServiceTests/Validators/AddRightsForUserValidatorTests.cs (hunk 14 above)
Comment (ifelseelif) [reply DigitalOffice/Tests/CheckRightsServiceTests/Validators/AddRightsForUserValidatorTests.cs]: renamed

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs (hunk 15 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: I do not understand how? As for me it's easier to understand

**Review by KostinP — APPROVED** [review]

File: DigitalOffice/Services/CheckRightsService/Database/Entities/DbRightUser.cs [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace LT.DigitalOffice.CheckRightsService.Database.Entities
+{
+    public class DbRightUser
+    {
+        public Guid UserId { get; set; }
+        public int RightId { get; set; }
+        public DbRight Right { get; set; }
+    }
```
Comment (KostinP) [inline DigitalOffice/Services/CheckRightsService/Database/Entities/DbRightUser.cs:12]: Add empty line.

**Review by lalagvana — APPROVED** [review]

**Review by djino97 — APPROVED** [review]

File: DigitalOffice/Services/CheckRightsService/Database/Entities/DbRightUser.cs (hunk 20 above)
Comment (ifelseelif) [reply DigitalOffice/Services/CheckRightsService/Database/Entities/DbRightUser.cs:12]: added

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs [hunk 21]
```diff
@@ -19,5 +22,35 @@ public List<DbRight> GetRightsList()
...
+                        UserId = request.UserId,
+                        Right = dbRight,
+                        RightId = rightId,
+                    });
+                }
+            }
+
+            dbContext.SaveChanges();
+
+            return true;
```
Comment (DarkFelas) [inline DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: remove, use **void**

**Review by DarkFelas — APPROVED** [review]

**Review by denis-kriv — APPROVED** [review]

**Comment by KostinP** [comment]
Add check in AccessValidator

**Review by KostinP — CHANGES_REQUESTED** [review]
I'm also fixing...

**Review by KostinP — APPROVED** [review]

**Pull request merged**